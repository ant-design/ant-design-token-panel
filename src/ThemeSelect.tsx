import type { FC, ReactNode } from 'react';
import type { ThemeConfig } from '@madccc/antd/es/config-provider/context';
import makeStyle from './utils/makeStyle';
import classNames from 'classnames';
import React, { useMemo } from 'react';
import { Button, Dropdown, Menu } from '@madccc/antd';
import { PlusOutlined, CloseOutlined } from '@ant-design/icons';

export type Theme = {
  name: string;
  key: string;
  theme: ThemeConfig;
  icon?: ReactNode;
  closable?: boolean;
  fixed?: boolean;
};

export type ThemeSelectProps = {
  onEnabledThemeChange: (themes: string[]) => void;
  onShownThemeChange: (themes: string[]) => void;
  enabledThemes: string[];
  shownThemes: string[];
  themes: Theme[];
};

const useStyle = makeStyle('ThemeSelect', (token) => ({
  '.previewer-theme-select': {
    padding: `${token.paddingXXS}px ${token.paddingXS}px`,
    borderRadius: 4,
    backgroundColor: 'rgba(0, 0, 0, 0.02)',
    height: token.controlHeight,
    display: 'flex',
    alignItems: 'center',
    overflow: 'hidden',

    '.ant-btn.previewer-theme-select-add-btn': {
      minWidth: 0,
      width: 16,
      height: 16,
      fontSize: 8,
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      marginLeft: token.marginSM,
    },

    '.previewer-theme-select-tag': {
      height: 22,
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center',
      boxSizing: 'border-box',
      borderRadius: 4,
      backgroundColor: token.colorBgComponent,
      border: `${token.lineWidth}px ${token.lineType} ${token.colorBorder}`,
      paddingInline: 10,
      marginRight: token.marginXS,
      fontSize: token.fontSizeSM,
      position: 'relative',
      cursor: 'pointer',
      transition: `all ${token.motionDurationMid}`,

      '&.previewer-theme-select-tag-active': {
        border: `${token.lineWidth}px ${token.lineType} ${token['blue-1']}`,
        backgroundColor: token['blue-1'],
        color: token.colorPrimary,

        '&::after': {
          content: '""',
          borderTopRightRadius: 2,
          position: 'absolute',
          right: 2,
          top: 2,
          width: 6,
          height: 6,
          background: `linear-gradient(to right top, transparent, transparent 50%, ${token.colorPrimary} 50%, ${token.colorPrimary} 100%)`,
        },
      },

      '.previewer-theme-select-tag-close-btn': {
        position: 'absolute',
        top: -2,
        right: -2,
        width: 12,
        height: 12,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        background: token.colorBgComponent,
        boxShadow:
          '0 2px 8px -2px rgba(0,0,0,0.05), 0 1px 4px -1px rgba(25,15,15,0.07), 0 0 1px 0 rgba(0,0,0,0.08)',
        borderRadius: '50%',
        opacity: 0,
        pointerEvents: 'none',
        zIndex: 2,

        '> .anticon': {
          fontSize: 6,
        },
      },

      '&:hover': {
        '.previewer-theme-select-tag-close-btn': {
          opacity: 1,
          pointerEvents: 'initial',
        },
      },
    },
  },

  '.previewer-theme-select-dropdown': {
    '.previewer-theme-select-dropdown-title': {
      '.ant-dropdown-menu-item-group-title': {
        fontSize: token.fontSizeSM,
        paddingBottom: token.padding,
        paddingTop: 10,
      },
    },
  },
}));

const ThemeSelect: FC<ThemeSelectProps> = (props) => {
  const {
    onEnabledThemeChange,
    onShownThemeChange,
    enabledThemes,
    shownThemes,
    themes,
  } = props;

  const [wrapSSR, hashId] = useStyle();

  const dropdownItems = useMemo(
    () => [
      {
        disabled: true,
        label: '添加主题即可预览',
        className: 'previewer-theme-select-dropdown-title',
        type: 'group',
        key: 'add-theme-to-preview',
      },
      ...themes
        .filter((theme) => !shownThemes.includes(theme.key))
        .map((theme) => ({
          icon: theme.icon,
          value: theme.key,
          label: theme.name,
          key: theme.key,
          onClick: () => {
            onShownThemeChange([...shownThemes, theme.key]);
          },
        })),
    ],
    [themes, shownThemes, onShownThemeChange],
  );

  const shownThemeEntities = useMemo(
    () => themes.filter((theme) => shownThemes.includes(theme.key)),
    [themes, shownThemes],
  );

  return wrapSSR(
    <div className={classNames('previewer-theme-select', hashId)}>
      {shownThemeEntities.map((theme) => (
        <span
          onClick={() => {
            if (theme.fixed) {
              return;
            }
            onEnabledThemeChange(
              enabledThemes.includes(theme.key)
                ? enabledThemes.filter((item) => item !== theme.key)
                : [...enabledThemes, theme.key],
            );
          }}
          key={theme.key}
          className={classNames('previewer-theme-select-tag', {
            'previewer-theme-select-tag-active': enabledThemes.includes(
              theme.key,
            ),
          })}
        >
          {theme.name}
          {theme.closable && (
            <span
              className="previewer-theme-select-tag-close-btn"
              onClick={(e) => {
                e.stopPropagation();
                onEnabledThemeChange(
                  enabledThemes.filter((item) => item !== theme.key),
                );
                onShownThemeChange(
                  shownThemes.filter((item) => item !== theme.key),
                );
              }}
            >
              <CloseOutlined />
            </span>
          )}
        </span>
      ))}
      <Dropdown
        placement="bottomRight"
        trigger={['click']}
        overlay={<Menu items={dropdownItems} />}
        overlayClassName={classNames('previewer-theme-select-dropdown', hashId)}
      >
        <Button
          type="primary"
          shape="circle"
          className="previewer-theme-select-add-btn"
          icon={<PlusOutlined />}
        />
      </Dropdown>
    </div>,
  );
};

export default ThemeSelect;
