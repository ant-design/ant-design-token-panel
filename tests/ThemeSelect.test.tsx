import { fireEvent, render, screen } from '@testing-library/react';
import ThemeSelect from '../src/ThemeSelect';
import React, { useState } from 'react';
import type { Theme } from '../src/interface';

const themes: Theme[] = [
  { name: 'Default Theme', key: 'default', config: {}, fixed: true },
  {
    name: 'Dark Theme',
    key: 'dark',
    config: {},
    closable: true,
  },
  {
    name: 'Compact Theme',
    key: 'compact',
    config: {},
    closable: true,
  },
];

const Demo = () => {
  const [shownThemes, setShownThemes] = useState<string[]>(['default', 'dark']);
  const [enabledThemes, setEnabledThemes] = useState<string[]>(['default']);

  return (
    <ThemeSelect
      enabledThemes={enabledThemes}
      shownThemes={shownThemes}
      themes={themes}
      onEnabledThemeChange={(value) => setEnabledThemes(value)}
      onShownThemeChange={(value) => setShownThemes(value)}
    />
  );
};

describe('ThemeSelect', () => {
  it('should render correctly', () => {
    const { container } = render(<Demo />, {
      container: document.body,
    });
    expect(container).toMatchSnapshot();
  });

  it('should show selected themes correctly', () => {
    const { container } = render(<Demo />, {
      container: document.body,
    });
    fireEvent.click(
      container.querySelector('.previewer-theme-select-add-btn')!,
    );
    expect(
      container.querySelector('.previewer-theme-select-add-btn')?.className,
    ).toContain('ant-dropdown-open');
    fireEvent.click(container.querySelector('li.ant-dropdown-menu-item')!);
    expect(container).toMatchSnapshot();
  });

  it('should toggle active theme correctly', () => {
    const { container } = render(<Demo />);
    expect(
      container.querySelectorAll('.previewer-theme-select-tag-active').length,
    ).toBe(1);
    fireEvent.click(
      container.querySelectorAll('.previewer-theme-select-tag')[1],
    );
    expect(
      container.querySelectorAll('.previewer-theme-select-tag-active').length,
    ).toBe(2);
    fireEvent.click(
      container.querySelectorAll('.previewer-theme-select-tag')[1],
    );
    expect(
      container.querySelectorAll('.previewer-theme-select-tag-active').length,
    ).toBe(1);
  });

  it('should support remove theme from tags', () => {
    const { container } = render(<Demo />);
    expect(
      container.querySelectorAll('.previewer-theme-select-tag').length,
    ).toBe(2);
    fireEvent.click(
      container.querySelector('.previewer-theme-select-tag-close-btn')!,
    );
    expect(
      container.querySelectorAll('.previewer-theme-select-tag').length,
    ).toBe(1);
  });

  it('should not toggle fixed theme', () => {
    const { container } = render(<Demo />);
    expect(
      container.querySelector('.previewer-theme-select-tag')?.children.length,
    ).toBe(0);
    expect(
      container.querySelectorAll('.previewer-theme-select-tag-active').length,
    ).toBe(1);
    fireEvent.click(container.querySelector('.previewer-theme-select-tag')!);
    expect(
      container.querySelectorAll('.previewer-theme-select-tag-active').length,
    ).toBe(1);
  });
});
