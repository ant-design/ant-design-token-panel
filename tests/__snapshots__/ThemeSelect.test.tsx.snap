// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSelect should render correctly 1`] = `
<body>
  <div
    class="previewer-theme-select"
  >
    <span
      class="previewer-theme-select-tag previewer-theme-select-tag-active"
    >
      Default Theme
    </span>
    <span
      class="previewer-theme-select-tag"
    >
      Dark Theme
      <span
        class="previewer-theme-select-tag-close-btn"
      >
        <span
          aria-label="close"
          class="anticon anticon-close"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </span>
    </span>
    <button
      class="ant-btn ant-btn-circle ant-btn-primary ant-btn-icon-only ant-dropdown-trigger previewer-theme-select-add-btn"
      type="button"
    >
      <span
        aria-label="plus"
        class="anticon anticon-plus"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <defs>
            <style />
          </defs>
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </span>
    </button>
  </div>
</body>
`;

exports[`ThemeSelect should show selected themes correctly 1`] = `
<body>
  <div
    class="previewer-theme-select"
  >
    <span
      class="previewer-theme-select-tag previewer-theme-select-tag-active"
    >
      Default Theme
    </span>
    <span
      class="previewer-theme-select-tag"
    >
      Dark Theme
      <span
        class="previewer-theme-select-tag-close-btn"
      >
        <span
          aria-label="close"
          class="anticon anticon-close"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </span>
    </span>
    <span
      class="previewer-theme-select-tag"
    >
      Compact Theme
      <span
        class="previewer-theme-select-tag-close-btn"
      >
        <span
          aria-label="close"
          class="anticon anticon-close"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="close"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
            />
          </svg>
        </span>
      </span>
    </span>
    <button
      class="ant-btn ant-btn-circle ant-btn-primary ant-btn-icon-only ant-dropdown-trigger previewer-theme-select-add-btn"
      type="button"
    >
      <span
        aria-label="plus"
        class="anticon anticon-plus"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="plus"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <defs>
            <style />
          </defs>
          <path
            d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
          />
          <path
            d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="ant-dropdown previewer-theme-select-dropdown ant-dropdown-hidden"
        style="pointer-events: none;"
      >
        <ul
          class="ant-dropdown-menu ant-dropdown-menu-root ant-dropdown-menu-vertical ant-dropdown-menu-light"
          data-menu-list="true"
          role="menu"
          tabindex="0"
        >
          <li
            class="ant-dropdown-menu-item-group previewer-theme-select-dropdown-title"
            disabled=""
          >
            <div
              class="ant-dropdown-menu-item-group-title"
              title="添加主题即可预览"
            >
              添加主题即可预览
            </div>
            <ul
              class="ant-dropdown-menu-item-group-list"
            />
          </li>
        </ul>
        <div
          aria-hidden="true"
          style="display: none;"
        />
      </div>
    </div>
  </div>
</body>
`;
